File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
public class Solution {
    public int maxProduct(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        // Initialize max, min, and result with the first element
        int maxProd = nums[0];
        int minProd = nums[0];
        int result = nums[0];

        // Traverse from the second element
        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];

            // If number is negative, swap maxProd and minProd
            if (num < 0) {
                int temp = maxProd;
                maxProd = minProd;
                minProd = temp;
            }

            // Update maxProd and minProd for current number
            maxProd = Math.max(num, maxProd * num);
            minProd = Math.min(num, minProd * num);

            // Update result
            result = Math.max(result, maxProd);
        }

        return result;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution(object):
    def maxProduct(self, nums):
        if not nums:
            return 0

        # Initialize max, min, and result with the first element
        max_prod = min_prod = result = nums[0]

        # Traverse from the second element
        for i in range(1, len(nums)):
            num = nums[i]

            # When num is negative, swap max_prod and min_prod
            if num < 0:
                max_prod, min_prod = min_prod, max_prod

            # Update max_prod and min_prod for current position
            max_prod = max(num, max_prod * num)
            min_prod = min(num, min_prod * num)

            # Update result
            result = max(result, max_prod)

        return result



        